src/components/chips/js/chipsController.js

```diff
@@ -0,0 +1,196 @@
+(function () {
+  'use strict';
+  angular
+      .module('material.components.chips')
+      .controller('MdChipsCtrl', MdChipsCtrl);
+
+
+
+  /**
+   * Controller for the MdChips component. Responsible for adding to and
+   * removing from the list of chips, marking chips as selected, and binding to
+   * the models of various input components.
+   *
+   * @param $mdUtil
+   * @param $mdConstant
+   * @param $log
+   * @ngInject
+   * @constructor
+   */
+  function MdChipsCtrl ($mdUtil, $mdConstant, $log) {
+    /** @type {Object} */
+    this.$mdConstant = $mdConstant;
+
+    /** @type {$log} */
+    this.$log = $log;
+
+    /** @type {angular.NgModelController} */
+    this.ngModelCtrl = null;
+
+    /** @type {Object} */
+    this.mdAutocompleteCtrl = null;
+
+    /** @type {Array.<Object>} */
+    this.items = [];
+
+    /** @type {number} */
+    this.selectedChip = -1;
+
+    /**
+     * Model used by the input element.
+     * @type {string}
+     */
+    this.chipBuffer = '';
+
+    /**
+     * Whether to use the mdChipAppend expression to transform the chip buffer
+     * before appending it to the list.
+     * @type {boolean}
+     */
+    this.useMdChipAppend = false;
+
+    /**
+     * Whether the Chip buffer is driven by an input element provided by the
+     * caller.
+     * @type {boolean}
+     */
+    this.hasInputElement = false;
+  }
+
+
+  /**
+   * Handles the keydown event on the input element: <enter> appends the
+   * buffer to the chip list, while backspace removes the last chip in the list
+   * if the current buffer is empty.
+   * @param event
+   */
+  MdChipsCtrl.prototype.defaultInputKeydown = function(event) {
```

Comment:
I've written a lot of Google-Closure js, where Class.prototype.method = ... is the rule. Not sure how strongly we're adhering to a styleguide here in Angular (and if there is a written guide in the first place)... I've seen 2 or 3 different ways of defining class methods in this repo now. The style I've used here is just what I'm used to and so it looks the best to me. Ultimately, my preference is towards whatever is the most readable, so if you can read my code, and I don't have to change it, I'm not going to :)